gym/dev_wrappers/time_aware_observation.py

```diff
@@ -0,0 +1,105 @@
+"""Wrapper for adding time aware observations to environment observation."""
+from collections import OrderedDict
+
+import jumpy as jp
+
+import gym
+import gym.spaces as spaces
+from gym.core import ActType, ObsType
+from gym.spaces import Box, Dict
+
+
+class TimeAwareObservationV0(gym.ObservationWrapper):
+    """Augment the observation with the current time step in the episode.
+
+    Example:
+        >>> import gym
+        >>> env = gym.make('CartPole-v1')
+        >>> env = TimeAwareObservationV0(env)
+        >>> env.observation_space
+        Dict(obs: Box([-4.8000002e+00 -3.4028235e+38 -4.1887903e-01 -3.4028235e+38], [4.8000002e+00 3.4028235e+38 4.1887903e-01 3.4028235e+38], (4,), float32), time: Box(0.0, inf, (1,), float32))
+        >>> _ = env.reset()
+        >>> env.step(env.action_space.sample())[0]
+        OrderedDict([('time', array([1.])),
+        ...  ('obs',
+        ...    array([ 0.02768888,  0.1745313 ,  0.03663293, -0.32239535], dtype=float32))])
+
+    Flatten observation space example:
+        >>> env = gym.make('CartPole-v1')
+        >>> env = TimeAwareObservationV0(env, flatten=True)
+        >>> env.observation_space
+        Box([-4.8000002e+00 -3.4028235e+38 -4.1887903e-01 -3.4028235e+38  0.0000000e+00], [4.8000002e+00 3.4028235e+38 4.1887903e-01 3.4028235e+38 inf], (5,), float32)
+        >>> _ = env.reset()
+        >>> env.step(env.action_space.sample())[0]
+        array([-0.01232257,  0.19335455, -0.02244143, -0.32388705,  1. ], dtype=float32)
+    """
+
+    def __init__(self, env: gym.Env, flatten=False):
+        """Initialize :class:`TimeAwareObservationV0`.
+
+        Args:
+            env: The environment to apply the wrapper
+            flatten: Flatten the observation to a `Box` of a single dimension
+        """
+        super().__init__(env)
+        self.flatten = flatten
+        self.num_envs = getattr(env, "num_envs", 1)
+
+        self.time_aware_observation_space = Dict(
+            obs=env.observation_space, time=Box(0, jp.inf, (self.num_envs,))
+        )
+
+        if self.flatten:
+            self.observation_space = spaces.flatten_space(
+                self.time_aware_observation_space
+            )
+        else:
+            self.observation_space = self.time_aware_observation_space
+
+    def observation(self, observation: ObsType):
+        """Adds to the observation with the current time step.
+
+        Args:
+            observation: The observation to add the time step to
+
+        Returns:
+            The observation with the time step appended to
+        """
+        observation = OrderedDict(obs=observation, time=self.t)
```

Comment:
why not use the remaining time (as suggested by https://arxiv.org/abs/1712.00378 or done in https://sb3-contrib.readthedocs.io/en/master/common/wrappers.html#timefeaturewrapper)
unbounded observation sounds like a bad idea.

Reply:
I'm on board with that, either remaining time, or normalized by the max time.

Reply:
Have another parameter, normalise time that is by default true? Otherwise returns the integer remaining time steps?